node/modules/storageminer.go

```diff
@@ -357,76 +356,187 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
+		case errors.Is(err, api.ErrF3ParticipationTicketExpired):
+			log.Warnw("F3 participation ticket expired while attempting to participate. Acquiring a new ticket.", "attempts", p.backoff.Attempt(), "err", err)
+			return api.F3ParticipationLease{}, false, nil
+		case errors.Is(err, api.ErrF3ParticipationTicketInvalid):
+			log.Errorw("F3 participation ticket is not valid. Acquiring a new ticket after backoff.", "backoff", p.backoff.Duration(), "attempts", p.backoff.Attempt(), "err", err)
+			p.backOff(ctx)
+			return api.F3ParticipationLease{}, false, nil
+		case errors.Is(err, api.ErrF3ParticipationIssuerMismatch):
+			log.Warnw("Node is not the issuer of F3 participation ticket. Miner maybe load-balancing or node has changed. Retrying F3 participation after backoff.", "backoff", p.backoff.Duration(), "err", err)
+			p.backOff(ctx)
+			log.Debugw("Reattempting F3 participation with the same ticket.", "attempts", p.backoff.Attempt())
+			continue
+		case err != nil:
+			log.Errorw("Unexpected error while attempting F3 participation. Retrying after backoff", "backoff", p.backoff.Duration(), "attempts", p.backoff.Attempt(), "err", err)
+			p.backOff(ctx)
+			continue
+		default:
+			log.Infow("Successfully acquired F3 participation lease.", "issuer", lease.Issuer, "expiry", lease.ValidityTerm)
+			p.previousTicket = ticket
+			return lease, true, nil
+		}
+	}
+	return api.F3ParticipationLease{}, false, ctx.Err()
+}
 
-			if ctx.Err() != nil {
-				return
-			}
-			if errors.Is(err, full.ErrF3Disabled) {
-				log.Errorf("Cannot participate in F3 as it is disabled: %+v", err)
-				return
-			}
-			if err != nil {
-				log.Errorf("while starting to participate in F3: %+v", err)
-				// use exponential backoff to avoid hotloop
-				timer.Reset(b.Duration())
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
-			}
-			if !ok {
-				log.Errorf("lotus node refused our lease, are you loadbalancing or did the miner just restart?")
-
-				sleepFor := b.Duration()
-				if d := time.Until(oldLease); d > 0 && d < sleepFor {
-					sleepFor = d
-				}
-				timer.Reset(sleepFor)
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
+func (p *f3Participator) awaitLeaseExpiry(ctx context.Context, lease api.F3ParticipationLease) error {
+	p.backoff.Reset()
+	for ctx.Err() == nil {
+		switch progress, err := p.node.F3GetProgress(ctx); {
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot await F3 participation lease expiry as F3 is disabled.", "err", err)
+			return xerrors.Errorf("awaiting F3 participation lease expiry: %w", err)
+		case err != nil:
+			if p.backoff.Attempt() > float64(p.maxCheckProgressAttempts) {
+				log.Errorw("Too many failures while attempting to check F3  progress. Restarting participation.", "attempts", p.backoff.Attempt(), "err", err)
+				return nil
```

Comment:
returning nil?

Reply:
Yes, this is intentional to abandon current lease and start participation from scratch. I chose the [max attempts](https://github.com/filecoin-project/lotus/blob/3d12ca5c53c5df74bb9daa0ea9be27b283e2469c/node/modules/storageminer.go#L503-L512) value such that we can be confident the abandoned lease is expired when the participation is restarted with a new lease.

The rationale for restarting participation to begin with is that this error can happen if lotus node is down due to operational issues: maybe it is being restarted etc. Returning error here means the miner will drop out of participation entirely until it is restarted. which I believe is not what we want. Unless I have missed something?

Reply:
This code sketches me out.

1. We should never make decisions based on the instance the lotus node we _happen_ to be talking to is at.
2. We should never rely on "enough time".

Instead, we should always:

1. Keep track of the last instance we agreed to participate in.
2. Make sure any lotus node we try to renew the ticket on has that information (e.g., has the old ticket).

Reply:
Ah, I see. We'll _always_ keep the previous ticket anyways. In that case, this seems totally reasonable.